src/relations/opensearch.py

```diff
@@ -0,0 +1,222 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Defines opensearch relation event handling methods."""
+
+import logging
+import re
+
+import requests
+from charms.data_platform_libs.v0.data_interfaces import IndexCreatedEvent
+from ops import framework
+from ops.model import BlockedStatus, WaitingStatus
+from ops.pebble import ExecError
+from requests.auth import HTTPBasicAuth
+
+from literals import HEADERS, JAVA_HOME, OPENSEARCH_SCHEMA
+from utils import log_event_handler
+
+logger = logging.getLogger(__name__)
+
+
+class OpensearchRelationHandler(framework.Object):
+    """Client for ranger:postgresql relations.
+
+    Attributes:
+        INDEX_NAME: the opensearch index name.
+        CERTIFICATE_NAME: the name of the opensearch certificate
+    """
+
+    INDEX_NAME = "ranger_audits"
+    CERTIFICATE_NAME = "opensearch-ca"
+
+    def __init__(self, charm, relation_name="opensearch"):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+            relation_name: The name of the relation.
+        """
+        self.relation_name = relation_name
+        self.charm = charm
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            self.charm.opensearch_relation.on.index_created,
+            self._on_index_created,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken,
+            self._on_relation_broken,
+        )
+
+    @log_event_handler(logger)
+    def _on_index_created(self, event: IndexCreatedEvent) -> None:
+        """Handle openserach relation changed events.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.config["charm-function"] == "admin":
```

Comment:
Should this check live in the `_validate` method? I'm not sure if this will run only once upon relating or every time the charm update status hook is called. If it is added to the `_validate` method, we know that the charm will always be in a blocked state if this condition is met.